cpp/src/arrow/util/ubsan.h

```diff
@@ -52,20 +52,42 @@ inline T* MakeNonNull(T* maybe_null = NULLPTR) {
   return const_cast<T*>(reinterpret_cast<const T*>(&internal::kNonNullFiller));
 }
 
+// SafeLoadAs with a specific number of bytes
+// The returned bytes whose position is greater than `num_bytes` will be initialized to 0
+template <typename T>
+inline typename std::enable_if<std::is_trivial<T>::value, T>::type SafeLoadAs(
+    const uint8_t* unaligned, uint32_t num_bytes) {
```

Comment:
Hmm, looking at the tests, I have a question: what is the desired semantics on big-endian machine? It seems this would happily return a value with only high bits set. Shouldn't it be the converse (only low bits set)?

Reply:
Agreed that it doesn't look like a generally useful function.
IMHO, directly write `r = 0; std::memcpy(&r, ...)` at the caller side is better, a function is not necessary.

Reply:
This method is only used in the hash code. So I agree to use `memcpy` directly in the hash function side.

Reply:
@pitrou , @cyb70289 , now the `memcpy` is used at the caller directly, and the added `SafeLoad` is removed